185708758 hmis api synchronization

ttoomey opened the pull request:
## Description
https://www.pivotaltracker.com/story/show/185708758
https://github.com/greenriver/hmis-frontend/pull/470

Proactively handle conflicts that could arise from multiple people acting on shared records. Adds rails optimistic locking to some HMIS HUD client, customAssessment, and enrollment models to protect us conflicting mutations. 
* Conflicts result in an error message shown to the client instructing them to reload the page
* optimistic locking is disabled for the warehouse models to prevent unintended side-effects
* This is not intended to be comprehensive and more work is needed, perhaps expanding locking to other record types

Other changes:
* lock around the enrollment on submit_assessment, enrollment id is now required
* wrap submit_form mutation in transaction to ensure updates are atomic
* submit_household_assessment takes an array of {id, lockVersion} inputs rather than just the ID

## Testing / QA
Intake / Exit Assessment:
- open two tabs with the same assessment
- click "save changes" in each tab. One of the tabs should succeed and the other one should display an error message

Delete Assessment:
- open two tabs with the same assessment
- update the assessment in one tab, in the other tab attempt to delete the assessment, the delete should fail

Update Client Details:
- open two tabs with the same client
- click "save changes" in each tab. One of the tabs should succeed and the other one should display an error message
- click "submit" in each tab. One of the tabs should succeed and the other one should display an error message

Delete Client:
- open two tabs with the same client
- update the client details in one tab, in the other tab attempt to delete the client, the delete should fail

## Type of change
- [x] Bug fix


## Checklist before requesting review
- [x] I have performed a self-review of my code
- [x] I have run the code that is being changed under ideal conditions, and it doesn't fail
- [x] My code includes comments and/or descriptive variable names to help other engineers understand the intent (or not applicable)
- [x] My code follows the style guidelines of this project (rubocop)
- [x] I have updated the documentation (or not applicable)
- [x] If it's not obvious how to test this change, I have provided testing instructions in this PR or the related issue

ttoomey committed:
185708758-hmis-api-synchronization

ttoomey committed:
Merge branch 'release-86' into 185708758-hmis-api-synchronization

ttoomey committed:
185708758-hmis-api-synchronization

ttoomey committed:
185708758-hmis-api-synchronization

ttoomey committed:
185708758-hmis-api-synchronization

ttoomey committed:
Merge branch 'release-86' into 185708758-hmis-api-synchronization

ttoomey committed:
185708758-hmis-api-synchronization

comments, bugfix

gigxz commented on drivers/hmis/app/graphql/mutations/submit_form.rb:
```diff
@@ -9,21 +9,32 @@ class SubmitForm < BaseMutation
       definition = Hmis::Form::Definition.find_by(id: input.form_definition_id)
       raise HmisErrors::ApiError, 'Form Definition not found' unless definition.present?
 
+      # FIXME: locking around the form definition isn't right. Probably want to lock on record or record.owner
```
Agreed. I don't think there is any need to lock the definition, we're not updating it at all.

ttoomey replied:
The lock here is intended to prevent submissions for the same definition from running in parallel. The definition is convenient for this since it's guaranteed to exist and will be the same across mutations; it's essentially functioning as an advisory lock here so I don't think it matters that the definition isn't updated. But yes, it is a bit of a hack.

Perhaps I should build something that infers the correct record to lock on from the input or we could use an actual advisory lock.

gigxz replied:
> The lock here is intended to prevent submissions for the same definition from running in parallel.

We definitely don't want that. That would mean that no two intake assessments cross the entire system could be submitted at the same time. Or, in the case for SubmitForm, it would prevent 2 services from being recorded at the same time (for example).

ttoomey replied:
Per discussion on slack - the definition lock shouldn't have much practical impact at our current small scale. However it feels like a hack so I'll replace with simple tx for now; we can rely on optimistic locking for the record types that support it. If integrity issues surface we can re-examine this

gigxz commented on drivers/hmis/app/graphql/types/hmis_schema/assessment_input.rb:
```diff
@@ -16,8 +16,18 @@ class HmisSchema::AssessmentInput < Types::BaseInputObject
+    # FIXME: this method probably doesn't belong on the form input class
     def find_or_create_assessment
       form_definition = Hmis::Form::Definition.find_by(id: form_definition_id)
+      # FIXME: locking around the form definition isn't right. Probably want to lock enrollment or assessment
```
agreed

gigxz commented on drivers/hmis/app/graphql/types/hmis_schema/assessment_input.rb:
```diff
@@ -16,8 +16,18 @@ class HmisSchema::AssessmentInput < Types::BaseInputObject
     argument :confirmed, Boolean, 'Whether warnings have been confirmed', required: false
     argument :validate_only, Boolean, 'Validate assessment but don\'t submit it', required: false
 
+    # FIXME: this method probably doesn't belong on the form input class
```
Yea this should be refactored, it was put here as a quick way for `SaveAssessment` and `SubmitAssessment` to share some setup logic since they have the same input type

gigxz commented on drivers/hmis/lib/hmis_errors/api_error.rb:
```diff
@@ -8,6 +8,7 @@ module HmisErrors
   class ApiError < StandardError
     attr_reader :display_message
     INTERNAL_ERROR_DISPLAY_MESSAGE = 'An error occurred'.freeze
+    STALE_OBJECT_ERROR = "We couldn't save your changes because someone else made updates to the same data while you were working on it. Please refresh the page and try saving your changes again. If the issue persists, contact support for assistance.".freeze
```
"we" feels slightly off to me? but dont feel strongly about it
```suggestion
    STALE_OBJECT_ERROR = "Your changes couldn't be saved because someone else made updates to the same data while you were working on it. Please refresh the page and try saving your changes again. If the issue persists, contact support for assistance.".freeze
```

gigxz commented on drivers/hmis/app/graphql/mutations/submit_assessment.rb:
```diff
@@ -9,13 +9,15 @@ class SubmitAssessment < BaseMutation
     field :assessment, Types::HmisSchema::Assessment, null: true
 
-    def resolve(input:)
+    def resolve(input:, assessment_lock_version: nil)
```
How tricky do you think it would be to add this to `SubmitHouseholdAssessments`? I think it would need to accept an array of lock versions, one per assessment. I think we should probably do that, as those assessments should be locked too.

ttoomey replied:
It seems do-able, I'll give it a shot. I think we'd probably need to change it to an input array of [{assessment_id: , lock_version: }, ...]

ttoomey committed:
Merge branch 'release-88' into 185708758-hmis-api-synchronization

gigxz commented:
I think this looks good aside from the comments. I haven't had a chance to properly test it yet, though. Adding  @eanders as a second reviewer.

ttoomey committed:
Merge branch 'release-88' into 185708758-hmis-api-synchronization

gigxz commented:
@ttoomey I'm getting the `Attempted to touch a stale object` every time I try to submit an assessment, even if I don't have any other tabs doing things

ttoomey committed:
185708758-hmis-api-synchronization

bugfix

observed behavior:
When version is NULL, a no-op save followed by touch raises stale object error.

not sure why, using a non-null value for lock-version seems to fix it

ttoomey committed:
185708758-hmis-api-synchronization

misc fixes, version household submit

ttoomey committed:
185708758-hmis-api-synchronization

tidy

ttoomey commented:
> @ttoomey I'm getting the `Attempted to touch a stale object` every time I try to submit an assessment, even if I don't have any other tabs doing things

Thanks @gigxz - I think I tracked this down but not 100% sure. It seems to be related to updates on  existing records where lock_version was NULL. Updated the migration to use a default value of 0 and I can't reproduce it anymore. You might need to redo this migration if you have it locally:

`db:migrate:redo:warehouse VERSION=20230913042115`

gigxz commented:
Thanks @ttoomey! That part is working now.
For submitting household assessments, is the frontend PR not up to date?
<img width="951" alt="Screenshot 2023-09-22 at 9 48 54 AM" src="https://github.com/greenriver/hmis-warehouse/assets/5333982/3566850d-0e22-4217-a06a-9edce4a0cd83">

ttoomey commented:
> Thanks @ttoomey! That part is working now. For submitting household assessments, is the frontend PR not up to date? 

Great! Sorry I forgot to commit the FE changes. Just pushed those up a few minutes ago

ttoomey committed:
Merge branch 'release-88' into 185708758-hmis-api-synchronization

ttoomey committed:
185708758-hmis-api-synchronization

test fixes

ttoomey commented:
@eanders do you have any concerns with adding optimistic locking to the hud tables? It is disabled on the warehouse models

gigxz commented:
Looks like a couple tests need to be updated. After that, and if Elliot is good with it, feel free to merge!

eanders commented:
We discussed optimistic locking and decided that if it's going to break on the warehouse side it will be pretty obvious, so we should give it a try.

ttoomey committed:
185708758-hmis-api-synchronization

test fixes
